### PR TITLE
New simulation time series plugin

### PR DESCRIPTION
**Pull request by jorgenherje**
Create new simulation time series plugin `SimulationTimeSeries`

Based on the functionality in original `ReservoirSimulationTimeSeries` plugin, with additional functionality requests listed in following issue: https://github.com/equinor/webviz-subsurface/issues/698.

This PR includes and closes parts of the listed requests:
- Show multiple delta ensembles - create delta ensembles
- Stacking possibility - subplot per vector and subplot per ensemble

---
### Review feedback
Please give feedback on following points:
- [ ] Naming of objects in `_property_serialization`-folder
- [ ] Naming of objects in `types`-folder
- [ ] Can utility functions in `utils`-folder be handy for others and placed in `webviz-subsurface/_utils`-folder?
- [ ] Consider moving ProviderSet to `webviz_subsurface/_utils`-folder. Could be reusable, and additional functionality can be added based on needs.

---

### Summary
The work includes structuring and refactoring of code for enhanced plugin implementation. Goal is to have code separated on functionality (callbacks, layout, business logic etc.), convert dash property to strongly typed arguments for "business logic". Clarifies intended usage and testability, preventing loosely typed arguments resulting in passing incorrect data. 

Utilize best practice example in [`webviz-plugin-boilerplate`](https://github.com/equinor/webviz-plugin-boilerplate).

---

### Main points
- Use ensemble summary data providers - https://github.com/equinor/webviz-subsurface/pull/721
    - Allows resampling frequency selectable
- Refactor and structure code for readability
    - Separate code dependent - layout, callback and business logic used in callback
    - Improve testability unit tests
    - Use boilerplate best practice plugin - [`webviz-plugin-boilerplate`](https://github.com/equinor/webviz-plugin-boilerplate)
    - Introduce strongly typed functionality for enhanced readability and more strict interfaces
- Show multiple `Delta Ensemble`:
   - Create `Delta Ensemble` 
   - `Delta Ensemble` as selectable ensemble
- Group by:
   - `vectors` - subplots per selected vector
   - `ensemble`  - subplots per selected ensemble

---

### Todo

- Unit testing of "business logic"

### CODE REVIEW

File: webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# TODO: Rename?
+# "Possessor"
+# "Accessor"
+# Suggestions:
+# - DerivedEnsembleDataAccessor
+# - DerivedEnsembleData
+# - DerivedEnsembleVectors
+# - DerivedEnsembleVectorsProvider
+# - DerivedVectorsProvider
+# - DerivedVectorsAccessor
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py:16]: Should we ask the team and decide on name? Or just go for one of them?

File: webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor_impl.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        cumulative_vector_names = [
+            get_cumulative_vector_name(elm)
+            for elm in self._interval_and_average_vectors
+            if is_interval_or_average_vector(elm)
+        ]
+        cumulative_vector_names = list(sorted(set(cumulative_vector_names)))
+
+        # TODO: Fetch vectors df with correct sampling and perform calculation.
+        # Ensure valid sampling and ensure correct AVG_ calculation (unit/day)
+        # calculation, i.e num days between sampling points
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor_impl.py:126]: Issue?

**Review by anders-kiaer — CHANGES_REQUESTED** [review]
Great work @jorgenherje 👏🚀 Looks nice!

I like that you also apparently has laid the ground for reusing parts of the functions in other plugins as well going forward.

There are some `TODO`s in the code (I've only commented on a couple of them). Should we stick with them in code only, or should we make issue of them as well as reminders?

Besides that I have only a couple of minor comments, then this is ready to merge IMO when we see green light towards example data.

File: webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py (hunk 1 above)
Comment (jorgenherje) [reply webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py:16]: Would be nice to get some input on the naming, as it feels long. Basically it should describe what it does - accessing vector data, both regular and calculated, for an ensemble (single or delta).

Perhaps the name is good enough as it is?

File: webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor_impl.py (hunk 2 above)
Comment (jorgenherje) [reply webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor_impl.py:126]: Should rather remove the comment, as calculation of AVG_ and INTVL_ now follows the same resampling as providers. Previously the calculation had its own frequency setting (as the original time series plugin). As AVG_ is [unit/day], the coming work would be to handle calculation if raw data is included - due to inconsistent sampling frequency and frequency which can be higher than one sample per day.

File: webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py (hunk 1 above)
Comment (anders-kiaer) [reply webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py:16]: The class name itself is perhaps not being written out in full in so many places (only where initialized), so current (long) name is perhaps good enough 👍

Comment (jorgenherje) [reply webviz_subsurface/plugins/_simulation_time_series/types/derived_ensemble_vectors_accessor.py:16]: Yes, I feel the other possible name would be `DerivedVectorsAccessor`, basically dropping "Ensemble" from the naming

File: webviz_subsurface/_utils/statistics_plotting.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    """
+
+    # TODO:
+    # - Rename mid percentile, find better name?
+    # - Consider to replace all lines with List[LineData], where each free line must be
+    # named and provided data.
+    # - Can then be used for individual realization plots as well?
+    # - One suggestion: Create base class with: samples: list, free_lines: List[LineData]
+    # and inherit for "StatisticsData". Base class can be utilized for realization plots?
+
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply webviz_subsurface/_utils/statistics_plotting.py:45]: Issue created: https://github.com/equinor/webviz-subsurface/issues/887

File: webviz_subsurface/_utils/statistics_plotting.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+def validate_statistics_data(data: StatisticsData) -> None:
+    """
+    Validation of statistics data
+
+    Ensure equal length of all statistical data lists and x-axis data list
+
+    Raise ValueError if lengths are unequal
+    """
+    if len(data.samples) <= 0:
```
[the remark this answers is not in the document]
Comment (jorgenherje) [reply webviz_subsurface/_utils/statistics_plotting.py:63]: Fixed.

NB: Applied to `fanchart_plotting.py` as well

**Review by anders-kiaer — APPROVED** [review]

**Pull request merged**